### PR TITLE
Add symmath to openedx-calc

### LINKED ISSUE #37: Move edx-platform/common/lib/symmath/symmath here
`common/lib/symmath` is one of six directories of `edx-platform` that are packaged in a very strange way. `common/lib/symmath` contains Symbolic math utils, used by untrusted scoring code. The code in this directory is not imported from the root of the directory as this directory is structured as an individual local Python projects, with its own `setup.py` and installed into the greater `edx-platform` via `requirements/edx/local.txt`.

For more context see [BOM-2579](https://openedx.atlassian.net/browse/BOM-2579)

This issue is part of [BOM-2581](https://openedx.atlassian.net/browse/BOM-2581).

### PR DESCRIPTION
**Pull request by Carlos-Muniz**
closes #37 

`symmath` is leaving `edx-platform` and moving to `openedx-calc`. As our first step we move `symmath` to `openedx-calc`. We update the `requirements`, test config files, the `readme`, the `setup.py`, and move all the tests to its own directory.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,6 +12,14 @@ clean:  ## Remove all build artifacts
...
 	tox
 
+requirements: ## install development environment requirements
+	pip install -r requirements/pip.txt
+	pip install -r requirements/pip_tools.txt
+	pip install -r requirements/base.txt
+	pip install -r requirements/test.txt
+	pip install -r requirements/tox.txt
+	pip install -r requirements/ci.txt
+
```
Comment (kdmccormick) [inline Makefile:22]: Generally we use the `make requirements` rule to install the requirements that a dev would need to work on the project. For this project, I think that's just `pip.txt`, `pip_tools.txt`, and `test.txt`. Note that `test.txt` is a superset of `base.txt`.

File: README.rst [hunk 2]
```diff
@@ -1,7 +1,7 @@
 openedx-calc
 ============
 
-A helper library for mathematical calculations, used by the `edx-platform`_.
+A helper library for mathematical calculations and symbolic mathematics, used by the `edx-platform`_.
```
Comment (kdmccormick) [inline README.rst:4]: Thanks for updating the README, and in general thanks for you thoroughness in this PR 💯

File: requirements/base.txt [hunk 3]
```diff
@@ -1,14 +1,22 @@
 #
-# This file is autogenerated by pip-compile
+# This file is autogenerated by pip-compile with python 3.8
```
Comment (kdmccormick) [inline requirements/base.txt:2]: Nice work splitting the PR into commits; made it really easy to review. The requirements upgrade especially is nice to have separate from the other changes.

**Review by kdmccormick — CHANGES_REQUESTED** [review]
Code changes looks solid 👍🏻 

Two issues with the tests though:

1. I don't think linting is being run. Both the `py38` and `quality` test contexts run unit tests; neither contexts run quality analysis. I think there's a disconnect in the github action -> tox machinery. Unfortunately it looks like this was the case before you opened your PR :(
2. I think `test_symmath_check_different_symbols` is silently failing

```
2022-01-13T20:39:52.0945218Z test_symmath_check_different_symbols (tests.test_symmath_check.SymmathCheckTest) ... Err 'lxml.etree._Element' object has no attribute 'decode' while preprocessing; expr=<math xmlns="http://www.w3.org/1998/Math/MathML">
2022-01-13T20:39:52.0952420Z   <mstyle displaystyle="true">
2022-01-13T20:39:52.0952888Z   <mrow>
2022-01-13T20:39:52.0953227Z     <mi>x</mi>
2022-01-13T20:39:52.0953469Z     <mo>+</mo>
2022-01-13T20:39:52.0953758Z     <mi>y</mi>
2022-01-13T20:39:52.0953991Z   </mrow>
2022-01-13T20:39:52.0954305Z </mstyle>
2022-01-13T20:39:52.0954559Z </math>
2022-01-13T20:39:52.0955459Z Error evaluating expression 'x+y' as a valid equation
2022-01-13T20:39:52.0955869Z Traceback (most recent call last):
2022-01-13T20:39:52.0964356Z   File "/home/runner/work/openedx-calc/openedx-calc/symmath/symmath_check.py", line 270, in symmath_check
2022-01-13T20:39:52.0965117Z     fsym = f.sympy
2022-01-13T20:39:52.0965849Z   File "/home/runner/work/openedx-calc/openedx-calc/symmath/formula.py", line 468, in make_sympy
2022-01-13T20:39:52.0966297Z     self.the_sympy = self.make_sympy(xml[0])
2022-01-13T20:39:52.0966775Z   File "src/lxml/etree.pyx", line 1176, in lxml.etree._Element.__getitem__
2022-01-13T20:39:52.0967216Z IndexError: list index out of range
2022-01-13T20:39:52.0967531Z ok
2022-01-13T20:39:52.0996244Z test_symmath_check_equivalent_symbols (tests.test_symmath_check.SymmathCheckTest) ... ok
2022-01-13T20:39:52.9970904Z test_symmath_check_floats (tests.test_symmath_check.SymmathCheckTest) ... ok
2022-01-13T20:39:53.7788603Z test_symmath_check_integers (tests.test_symmath_check.SymmathCheckTest) ... ok
2022-01-13T20:39:53.7984221Z test_symmath_check_same_symbols (tests.test_symmath_check.SymmathCheckTest) ... ok
2022-01-13T20:39:53.7984503Z 
2022-01-13T20:39:53.7985274Z ----------------------------------------------------------------------
2022-01-13T20:39:53.7985902Z Ran 64 tests in 3.923s
2022-01-13T20:39:53.7986042Z 
2022-01-13T20:39:53.7986327Z OK
```

File: Makefile (hunk 1 above)
Comment (Carlos-Muniz) [reply Makefile]: Requirements have been edited.

**Comment by kdmccormick** [comment]
@Carlos-Muniz Two other thoughts that just came to mind:

Firstly, you will want to bump the version of this project, potentially the major version since this is a relatively large change.

Secondly: by default, edx-platform will upgrade to the latest version of all project every week [unless they are explicitly constrainted](https://github.com/openedx/edx-platform/blob/master/requirements/constraints.txt). So, edx-platform would inherit these changes within a week after merging this PR if you release a new version (which you should). To be honest, I don't know what happens if two versions of `symmath` (the edx-platform/common/lib version and this one) are both available to import!

So, in edx-platform, you may want to constrain openedx-calc to its current, symmath-less version before merging this PR. Then, once you're ready, you could delete the edx-platform/common/lib copy of `symmath` and upgrade to the latest version of openedx-calc all in one edx-platform PR.

**Comment by Carlos-Muniz** [comment]
> Code changes looks solid 👍🏻
> 
> Two issues with the tests though:
> 
> 1. I don't think linting is being run. Both the `py38` and `quality` test contexts run unit tests; neither contexts run quality analysis. I think there's a disconnect in the github action -> tox machinery. Unfortunately it looks like this was the case before you opened your PR :(
> 2. I think `test_symmath_check_different_symbols` is silently failing
> 
> ```
> 2022-01-13T20:39:52.0945218Z test_symmath_check_different_symbols (tests.test_symmath_check.SymmathCheckTest) ... Err 'lxml.etree._Element' object has no attribute 'decode' while preprocessing; expr=<math xmlns="http://www.w3.org/1998/Math/MathML">
> 2022-01-13T20:39:52.0952420Z   <mstyle displaystyle="true">
> 2022-01-13T20:39:52.0952888Z   <mrow>
> 2022-01-13T20:39:52.0953227Z     <mi>x</mi>
> 2022-01-13T20:39:52.0953469Z     <mo>+</mo>
> 2022-01-13T20:39:52.0953758Z     <mi>y</mi>
> 2022-01-13T20:39:52.0953991Z   </mrow>
> 2022-01-13T20:39:52.0954305Z </mstyle>
> 2022-01-13T20:39:52.0954559Z </math>
> 2022-01-13T20:39:52.0955459Z Error evaluating expression 'x+y' as a valid equation
> 2022-01-13T20:39:52.0955869Z Traceback (most recent call last):
> 2022-01-13T20:39:52.0964356Z   File "/home/runner/work/openedx-calc/openedx-calc/symmath/symmath_check.py", line 270, in symmath_check
> 2022-01-13T20:39:52.0965117Z     fsym = f.sympy
> 2022-01-13T20:39:52.0965849Z   File "/home/runner/work/openedx-calc/openedx-calc/symmath/formula.py", line 468, in make_sympy
> 2022-01-13T20:39:52.0966297Z     self.the_sympy = self.make_sympy(xml[0])
> 2022-01-13T20:39:52.0966775Z   File "src/lxml/etree.pyx", line 1176, in lxml.etree._Element.__getitem__
> 2022-01-13T20:39:52.0967216Z IndexError: list index out of range
> 2022-01-13T20:39:52.0967531Z ok
> 2022-01-13T20:39:52.0996244Z test_symmath_check_equivalent_symbols (tests.test_symmath_check.SymmathCheckTest) ... ok
> 2022-01-13T20:39:52.9970904Z test_symmath_check_floats (tests.test_symmath_check.SymmathCheckTest) ... ok
> 2022-01-13T20:39:53.7788603Z test_symmath_check_integers (tests.test_symmath_check.SymmathCheckTest) ... ok
> 2022-01-13T20:39:53.7984221Z test_symmath_check_same_symbols (tests.test_symmath_check.SymmathCheckTest) ... ok
> 2022-01-13T20:39:53.7984503Z 
> 2022-01-13T20:39:53.7985274Z ----------------------------------------------------------------------
> 2022-01-13T20:39:53.7985902Z Ran 64 tests in 3.923s
> 2022-01-13T20:39:53.7986042Z 
> 2022-01-13T20:39:53.7986327Z OK
> ```

I've looked into this issue more. From what I can tell, this test is failing because formula.make_sympy (in formula.py) does not work correctly when interpreting xml. The recursive implementation fails when it sees a tag it is not familiar with, in this case "mstyle".

**Comment by Carlos-Muniz** [comment]
A new issue #39 has been made to add more tests that evaluate the `dynamath` evaluation.

File: symmath/symmath_check.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+            msg += Markup('<p>You entered: {fans} (note that a numerical answer is expected)</p>').\
+                format(fans=to_latex(fans))
+            return {'ok': False, 'msg': msg}
+
+    if fexpect == fans:
+        msg += Markup('<p>You entered: {fans}</p>').format(fans=to_latex(fans))
+        return {'ok': True, 'msg': msg}
+
+    ###### PMathML input ######
+    # convert mathml answer to formula
```
Comment (kdmccormick) [inline symmath/symmath_check.py:257]: If I recall correctly, we found that everything below this line is poorly test-covered, probably broken, and potentially unused.

That's a great insight we reached that we don't want to lose! Could you leave a comment here? You could even link out to #39 .

File: tox.ini [hunk 5]
```diff
@@ -17,6 +17,7 @@ deps =
 	pycodestyle
 	pylint
 commands =
-	pycodestyle calc/__init__.py calc/tests/__init__.py
-	pylint calc/tests/__init__.py
+	pycodestyle calc/__init__.py
+	pycodestyle symmath/__init__.py
+	pylint tests/__init__.py
```
Comment (kdmccormick) [inline tox.ini:22]: Check out the logs of the `quality` PR check -- this isn't being run.

File: calc/__init__.py [hunk 6]
```diff
@@ -6,4 +6,4 @@
 
 from .calc import *
 
-__version__ = '2.0.1'
+__version__ = '3.0.0'
```
Comment (kdmccormick) [inline calc/__init__.py:9]: Thanks for the version bump.

Remember that upon merging this and [creating a new package release](https://github.com/openedx/openedx-calc/releases), edx-platform will automatically upgrade to the new version of this library, even though the old symmath hasn't been removed from edx-platform.  That may not be what you want. I recommend a [solution here](https://github.com/openedx/openedx-calc/pull/38#issuecomment-1013439750).

**Review by kdmccormick — CHANGES_REQUESTED** [review]
Looking pretty good; just a couple more issues.

File: symmath/symmath_check.py (hunk 4 above)
Comment (Carlos-Muniz) [reply symmath/symmath_check.py:257]: I think I understand what you want now. I've added the comment on line 258.

**Comment by kdmccormick** [comment]
@Carlos-Muniz this will not fix the linting violations, but it will make so `Python CI / Tests (ubuntu-20.04, 3.8, py38)` runs only unit tests and `Python CI / Tests (ubuntu-20.04, 3.8, quality)` runs only quality checks: https://github.com/openedx/openedx-calc/pull/40. Right now they're both running both, which is confusing.

**Comment by kdmccormick** [comment]
@Carlos-Muniz As for the new quality failures that we're seeing: no, you don't need to fix all of those, but we do need to get the build passing (without disabling quality checking altogether). Here's what I'd recommend:

For pycodestyle:
* Create a [`./setup.cfg` with a `pycodestyle` section](https://github.com/openedx/edx-platform/blob/bebe1e9a6be0e357053f59f1cf92e8d770d4c16f/setup.cfg#L22) and disable style rules that you think aren't helpful or are just beyond fixing at this point. For example, disabling `E501` (line too long) will eliminate a looot of those violations.

For pylint:
* You can use [edx-lint](https://github.com/openedx/edx-lint/) (will need to be added to this repo's testing requirements) to generate a [`./pylintrc` file](https://github.com/openedx/edx-ace/blob/master/pylintrc) for this repo, which will tell `pylint` to ignore certain violations. Similar to pycodestyle, you can choose to ignore violations that you think are unhelpful or are just too far gone to be worth fixing at this point.

File: tox.ini [hunk 7]
```diff
@@ -17,6 +17,6 @@ deps =
 	pycodestyle
 	pylint
 commands =
-	pycodestyle calc/__init__.py calc/tests/__init__.py
-	pylint calc/tests/__init__.py
+	pycodestyle calc symmath
+	pylint tests
```
Comment (kdmccormick) [inline tox.ini:21]: @Carlos-Muniz Ideally, we should be running both pycodestyle and pylint on all three root directories (calc, symmath, and tests).

I don't know why we were only running these checks on specific files before... I think the original author of this `tox.ini` may have been confused.

Comment (kdmccormick) [reply tox.ini:21]: Ok, turns out these checks were intentionally only being run on specific files: https://github.com/openedx/openedx-calc/pull/1#issuecomment-485952972. The original extractor of `openedx-calc` from edx-platform didn't want to fix all the pylint and pycodestyle violations ;)

I am OK leaving the existing sins in place (ie, not running quality checks on the code that was already here), but I'd like for the new symmath code & tests you are adding to be totally style-checked and linted, as they were in edx-platform.

File: tox.ini (hunk 5 above)
Comment (Carlos-Muniz) [reply tox.ini]: Quality tests are running

File: tox.ini (hunk 7 above)
Comment (Carlos-Muniz) [reply tox.ini]: I've run quality checks and removed all the lint.

File: symmath/formula.py [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+import six
+import sympy
+from lxml import etree
+from sympy import latex, sympify
+from sympy.physics.quantum.qubit import Qubit
+from sympy.physics.quantum.state import Ket
+from sympy.printing.latex import LatexPrinter
+from sympy.printing.str import StrPrinter
+
+# from openedx.core.djangolib.markup import HTML
```
Comment (kdmccormick) [inline symmath/formula.py:34]: Nit - unnecessary line
```suggestion
```

File: symmath/symmath_check.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+#
+# Takes in math expressions given as Presentation MathML (from ASCIIMathML), converts to Content MathML using SnuggleTeX
+
+
+import logging
+import traceback
+
+from markupsafe import escape, Markup
+
+# from openedx.core.djangolib.markup import HTML
```
Comment (kdmccormick) [inline symmath/symmath_check.py:18]:
```suggestion
```

**Review by kdmccormick — APPROVED** [review]
Nice work!

Two nits, but no need for another round of review IMO

File: symmath/formula.py (hunk 8 above)
Comment (Carlos-Muniz) [reply symmath/formula.py]: removed

File: symmath/symmath_check.py (hunk 9 above)
Comment (Carlos-Muniz) [reply symmath/symmath_check.py]: removed

**Pull request merged**